### PR TITLE
Include potential statusCode for next.js Error return

### PR DESCRIPTION
**Pull request by rgbkrk**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: `createError` from `micro` creates the errors with `statusCode` set: https://github.com/zeit/micro/blob/069f31fa506086cdbf0ce3364d16ab972e8cc200/lib/index.js

### CODE REVIEW

File: types/next/index.d.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ declare namespace next {
         /** Fetch Response object (client only) - from https://developer.mozilla.org/en-US/docs/Web/API/Response */
         jsonPageRes?: NodeResponse;
         /** Error object if any error is encountered during the rendering */
-        err?: Error;
+        err?: Error & {statusCode? : number};
```
Comment (mohsen1) [inline types/next/index.d.ts]: It looks like that the type of `err` is not `Error`. Guessing it's a `Response` object?

Comment (rgbkrk) [reply types/next/index.d.ts]: Within the next code base they use [this `createError` function from `micro`](https://github.com/zeit/micro/blob/069f31fa506086cdbf0ce3364d16ab972e8cc200/lib/index.js#L19-L26):

```
const createError = (code, message, original) => {
	const err = new Error(message);

	err.statusCode = code;
	err.originalError = original;

	return err;
};
```

for the errors, appending this `statusCode`. They then [suggest in the next docs when creating your own `_error.js` page](https://github.com/zeit/next.js#custom-error-handling) to rely on this `statusCode` property.

**Comment by joaovieira** [comment]
@rgbkrk you can attach anything to your Error. `statusCode` is just an example that Next.js suggests or uses in their default server implementation. Same applies for `originalError` apparently. I, for example, have a custom Express.js server and can attach anything to it.

Thus, I'd say this needs to be cast explicitly on your client, either to `any` or to your custom error type:

```ts
static getInitialProps(context: NextContext) {
  const { err, res } = context;
  const statusCode = res ? res.statusCode : err ? (err as any).statusCode : null;

  return { statusCode };
}
```

You can also override your `NextContext` type:

```ts
interface MyError extends Error {
  statusCode?: number;
  originalError?: Error;
}

interface MyNextContext extends NextContext {
  err?: MyError
}

static getInitialProps(context: NextContext) {
  const { err, res } = context;
  const statusCode = res ? res.statusCode : err ? err.statusCode : null;

  return { statusCode };
}
```

**Review by resir014 — CHANGES_REQUESTED** [review]
@rgbkrk @joaovieira Since the `err` property is only used in `_error.tsx` [as the documentation suggests](https://github.com/zeit/next.js#custom-error-handling), it might be better to limit the scope of this change to `error.d.ts` instead. I think we're better off doing this:

```tsx
// error.d.ts

export interface NextErrorContext extends NextContext {
  err: DefaultErrorIProps
}

export default class Error<P = {}> extends React.Component<P & DefaultErrorIProps> {
    static getInitialProps(context: NextErrorContext): Promise<DefaultErrorIProps> | DefaultErrorIProps;
}
```

**Comment by resir014** [comment]
@rgbkrk I added my change suggestion. My finger slipped so my message might have accidentally got cut short if you received your notifications via email. 😅

**Comment by joaovieira** [comment]
@resir014 I think we should be able to use `err.statusCode` in any page, page HOC, or _app. It's the same error. No?

File: types/next/index.d.ts (hunk 1 above)
Comment (joaovieira) [reply types/next/index.d.ts]: if we go with this approach I'd suggest including `originalError` for completeness as well

**Comment by resir014** [comment]
> I think we should be able to use `err.statusCode` in any page, page HOC, or _app. It's the same error, same context. No?

Ah, you're right, `err` is the same on every page. I stand corrected. @joaovieira

**Review by resir014 — APPROVED** [review]

**Comment by rgbkrk** [comment]
Thank you all so much for digging into this to find the truth!

**Comment by amcasey** [comment]
@joaovieira @mohsen1 Do you have any remaining concerns?

**Review by joaovieira — CHANGES_REQUESTED** [review]
@rgbkrk as you pointed out the default Next.js error also has an `originalError` prop. Could you add some tests as well?

**Comment by rgbkrk** [comment]
I'm going on paternity leave for a bit, so if anyone wants to pick up this PR (or even start a new one), I'd really appreciate it!

**Pull request closed without merging**